### PR TITLE
Introduce platform namespace

### PR DESCRIPTION
**Pull request by cmarcelo**
Introduce a new namespace 'platform' for functions that are unlikely to change after link time and would have no benefit from explicit indirection. This includes: `panic()`, `millis()`, `micros()`. The idea is that those basic utility functions could be provided as regular functions instead of virtual functions.

Attention: this PR contains multiple commits, each one has it's own detailed explanation.

### CODE REVIEW

**Comment by cmarcelo** [comment]
@tridge @lucasdemarchi could you take a look?

Other candidates for this are the functions in Util.h. If/when we feel that the header file is too big, we can split it functionally.

File: APMrover2/Log.cpp [hunk 1]
```diff
@@ -172,7 +172,7 @@ void Rover::Log_Write_Performance()
 {
     struct log_Performance pkt = {
         LOG_PACKET_HEADER_INIT(LOG_PERFORMANCE_MSG),
-        time_us         : hal.scheduler->micros64(),
+        time_us         : platform::micros64(),
         loop_time       : millis()- perf_mon_timer,
```
Comment (lucasdemarchi) [inline APMrover2/Log.cpp]: leftover?

**Comment by lucasdemarchi** [comment]
Just did a quick review... it's looking great. So, just with these changes we have a) faster and b) smaller code (~3KB, as per https://github.com/cmarcelo/ardupilot/commit/33fabf341e226be020b952c706c6bf5bb87eac66)

File: APMrover2/Log.cpp (hunk 1 above)
Comment (cmarcelo) [inline APMrover2/Log.cpp]: Vehicles that had their own `millis()` and `micros()` got them replaced with a `using platform::millis` and `using platform::micros`.

**Comment by tridge** [comment]
I'm happy with this, but I'd like to get the OK from Randy too

**Comment by rmackay9** [comment]
The general idea is fine with me although I'm not a huge fan of "platform".  It's quite a general purpose word that doesn't really seem to provide an indication of what should or shouldn't be included within it.

for me "util" or "clock" or something like that would be specific and useful.

**Comment by magicrub** [comment]
I agree, "platform" is no good. I'm in favor of util or global.

**Comment by lucasdemarchi** [comment]
platform is to mean the the OS/platform we are running on. Like Linux/px4, etc. Quite a common name in other projects. And, looking on what PX4 is doing now... it's using the same word :-)
https://github.com/PX4/Firmware/tree/master/src/platforms

In their case it's a platform separation, but without namespace.

Another option that comes to mind is "os", which is short, easy to type and almost covers the same meaning.  My preference is still platform though.

**Comment by magicrub** [comment]
Well I'm certainly more inclined to accept "platform" if there is a
precedent.

**Comment by tridge** [comment]
what about using HAL ?
so HAL::millis() and HAL::micros()
The advantage is that everyone developing for ardupilot already needs to know what HAL means

**Comment by cmarcelo** [comment]
@tridge `HAL` is the name of a class, so I rather avoid the conflict. What about just settling down with the existing namespace `AP_HAL` for now?

Note: I like `clock` in the context of the time related functions (not as namespace but as part of name), but thought that millis/micros was already too familiar to change in this batch.

**Comment by rmackay9** [comment]
I agree HAL would be confusing with people asking, what's the difference between "HAL" and "hal"?
if px4 and other projects are using "platform" then I'm ok with it.

**Comment by rmackay9** [comment]
Sorry, I take back my "ok".  I just think it's a backwards step in terms of naming and will confusion for users.  hal.scheduler->millis() makes sense.  platform::millis() doesn't to me at least.

**Comment by cmarcelo** [comment]
@rmackay9 would using the existing namespace `AP_HAL` acceptable? (e.g. `AP_HAL::millis()`)

**Comment by lucasdemarchi** [comment]
@rmackay9 what alternative do you prefer? Unless we will start from beginning with "split namespaces", the alternatives IMO are worse.

By split namespaces I mean:

```
clock::millis()
clock::micros()
sched::panic()
```

Or something like above. Then for each new function dependent on the platform we would check a namespace where it belongs.  Right now the alternatives for a single namespaces are:

```
platform::millis()
global::millis()
hal::millis()
HAL::millis()
AP_HAL::millis()
util::millis()
os::millis()
```

hal, HAL and util either shadow a variable or a class. There's AP_HAL. My grip with that one is that the functions are certainly not a **hardware** abstraction... they are abstracting the different functions from the OS. :-/. This deserves a proper quote: http://martinfowler.com/bliki/TwoHardThings.html

**Comment by rmackay9** [comment]
@cmarcelo @lucasdemarchi AP_HAL works for me, so does clock, sched and maybe even os.  I don't think we should shadow hal, HAL or util.

**Comment by rmackay9** [comment]
decided on the dev call that we can call it AP_HAL.

**Comment by cmarcelo** [comment]
Replaced by #3190.

**Pull request closed without merging**